### ISSUE TITLE
[BUG] Mixing `rmm::device_vector` and non-default stream is unsafe

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#2631]
**Describe the bug**

As identified in https://github.com/rapidsai/cudf/issues/2630 mixing the use of an `rmm::device_vector` with kernels executing on the non-default stream is dangerous. 

For example,

```
rmm::device_vector<int> v(100); // allocation happens on null stream
some_kernel<<<...., non_default_stream>>>(v.data().get()...) // execution occurs on non-null stream
```
In non-pool mode, the above is safe because allocation defaults to `cudaMalloc`, which is synchronous.

However, if RMM is used with the pool enabled (i.e., using CNMEM), the above example is broken. `cnmemMalloc` is _asynchronous_ wrt to the host, and therefore `v` is not guaranteed to be allocated before `some_kernel` is launched.  

**Expected behavior**
There needs to be a safe way to mix use of a RAII object like `rmm::device_vector` with non-default streams. 

**Additional context**
RMM should provide a `device_vector` like abstraction that allows a way to specify a stream for allocation.

**Comment by harrism** [comment]
If we update to the latest version of Thrust (by submoduling), rather than using the Thrust in the CTK, then we can easily provide a simple wrapper that takes a stream, since the latest Thrust has a device_vector that takes an allocator (rather than one that is just templated on an allocator).

**Comment by trevorsm7** [comment]
Should https://github.com/rapidsai/cudf/issues/1077 be taken care of before this then?

**Comment by harrism** [comment]
One concern is this would make a clone of RMM >6x larger than it currently  is (RMM clean clone is 5.2MB, thrust is 27MB).

**Comment by kkraus14** [comment]
We could get the latest version of thrust via CMake configuration so clone is unaffected but build time is increased a bit.

**Comment by harrism** [comment]
This should not have been closed by #5482 as that PR only addresses one module of many.
